config/m4/module.m4

```diff
@@ -0,0 +1,484 @@
+#
+# Usage: UCX_CHECK_MODULE([id], [name], [prefix], [pkgname],
+#                         [[doc], [doc-lib], [doc-static or '-' to skip], [doc-failed]],
+#                         [action-if-found], [action-if-not-found],
+#                         [[headers], [decls], [includes], [libs - space separated], [funcs], [otherlibs]],
+#                         [alt-dirs])
+#
+# Example:
+# m4_define([t1_arg], [[ucp/api/ucp.h, ucs/config/global_opts.h],
+#                      [UCP_FEATURE_TAG, ucp_init_version, ucs_global_opts_init],
+#                      [[#include <ucp/api/ucp.h>
+#                        #include <ucs/config/global_opts.h>]],
+#                      [ucp ucs], [ucp_tag_send_nb, ucs_global_opts_init], [-luct]])
+# 
+# UCX_CHECK_MODULE(t1, [name T1], UCX, ucx,
+#                  [[T1 library], [search T1 library in DIR], [Static T1 library], [Failed to detect T1 library]],
+#                  [AC_MSG_NOTICE(done)], [AC_MSG_NOTICE(failed)], [t1_arg])
+#
+AC_DEFUN([UCX_CHECK_MODULE],
+         [
+             # AS_VAR_PUSHDEF doesn't work here due to processed in autoconf.
+             # use m4_pushdef instead to instantiate id
+             m4_pushdef([mod_hlp], [$5])
+             AC_ARG_WITH([$1],
+                         [AS_HELP_STRING([--with-$1=(DIR)],
+                                         _CHECK_MODULE_MAKE_HELP_STRING([1], [mod_hlp], [ (default is guess).]))],
+                         [], [with_$1=guess])
+
+             AC_ARG_WITH([$1-libdir],
+                         [AS_HELP_STRING([--with-$1-libdir=(DIR)],
+                                         _CHECK_MODULE_MAKE_HELP_STRING([2], [mod_hlp], [.]))],
+                         [], [with_$1_libdir=""])
+
+             m4_if(m4_argn(3, $5), [-], [], [
+                 AC_ARG_WITH([$1-static],
+                             [AS_HELP_STRING([--with-$1-static=(DIR)],
+                                             _CHECK_MODULE_MAKE_HELP_STRING([3], [mod_hlp], [ (default is no).]))],
+                             [], [with_$1_static=no])])
+
+             # define 'local' variables
+             AS_VAR_PUSHDEF([mod_name], [$2])
+             AS_VAR_PUSHDEF([mod_prefix], [$3])
+             AS_VAR_PUSHDEF([mod_pkg], [$4])
+             m4_pushdef([mod_alt_dir], [$9])
+
+             _CHECK_MODULE_DEFVAR([$8])
+
+             AS_VAR_PUSHDEF([mod_with], [with_$1])
+             AS_VAR_PUSHDEF([mod_with_libdir], [with_$1_libdir])
+             AS_VAR_PUSHDEF([mod_with_static], [with_$1_static])
+             AS_VAR_PUSHDEF([mod_dir_inc], [include_$1])
+             AS_VAR_PUSHDEF([mod_dir_inc_flag], [include_flag_$1])
+             AS_VAR_PUSHDEF([mod_dir_lib], [lib_$1])
+             AS_VAR_PUSHDEF([mod_dir_flag], [lib_flag_$1])
+
+             mod_dir_inc=""
+             mod_dir_inc_flag=""
+             mod_dir_lib=""
+             mod_dir_flag=""
+
+             mod_progress=start
+
+             _CHECK_MODULE_RUN([$mod_progress], [start],
```

Comment:
i think there are too many steps in this macro, why need so many?
maybe add another helper macro like checking static lib?

Reply:
seems too many... i would expect something like:
1. check headers
2. check static libs
3. check dynamic libs

and for each such step: 1. search user specified dir, 2. try pkgconfig, 3. use default paths

Maybe we can express this structure with a more generic macro internal helper so it will be more "compact" code?

Reply:
every test has own structure: static lib is checked by pkgconfig first because it automatically detects for dependencies and if it failed - try to check library directly. dynamic library is detected by library first to provide compatibility with "legacy" way of module check and if failed - try to use pkgconfig. if user specify any directory then test must fail in case if directory inaccessible. if user didn't set any directory and module could not be found in default location then alternative directories should be checked and failures on such checks should be ignored. in case if module is detected using PKG_CHECK_MODULES macro then _CFLAGS and _LIBS variables should not be updated manually because it is updated by script. there are tons of combinations of tests and user input and adding any structure here just adds very nested conditions.